File: F.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
	int n, m;
    cin >> n >> m;
    vector<int> a(n);
    for (int i = 0; i < n; i++) 
    	cin >> a[i];
    int c, k;
    for (int i = 0; i < m; i++) {
    	cin >> c >> k;
        a[c - 1] += k;
    }
    
    int max1 = 0, max2 = 1, max3 = 2;
    
    if (a[max1] < a[max2]) swap(max1, max2);
    if (a[max2] < a[max3]) swap(max2, max3);
    if (a[max1] < a[max2]) swap(max1, max2);
    
    for (int i = 3; i < n; i++) {
    	if (a[i] > a[max1]) {
        	max3 = max2;
            max2 = max1;
            max1 = i;
        } else if (a[i] > a[max2]) {
        	max3 = max2;
            max2 = i;
        } else if (a[i] > a[max3]) {
        	max3 = i;
        }
    }
    cout << max1 + 1 << " " << max2 + 1 << " " << max3 + 1 << endl;
    return 0;
}

File: F.pas
var
    n, m, i, c, k, max1, max2, max3: longint;
    a: array[1..100000] of longint;

procedure swap(var a, b: longint);
var
    t: longint;
begin
    t := a;
    a := b;
    b := t;
end;

begin
    read(n, m);
    for i := 1 to n do
        read(a[i]);
    for i := 1 to m do begin
        read(c, k);
        a[c] := a[c] + k;
    end;
    
    max1 := 1;
    max2 := 2;
    max3 := 3;

    if a[max1] < a[max2] then
        swap(max1, max2);
    if a[max2] < a[max3] then
        swap(max2, max3);
    if a[max1] < a[max2] then
        swap(max1, max2);

    for i := 4 to n do
        if a[i] > a[max1] then begin
            max3 := max2;
            max2 := max1;
            max1 := i;
        end else if a[i] > a[max2] then begin
            max3 := max2;
            max2 := i;
        end else if a[i] > a[max3] then
            max3 := i;
    writeln(max1, ' ', max2, ' ', max3);
end.
